File: src/sql/postgres/PostgresSQLStatement.zig
const PostgresSQLStatement = @This();
const RefCount = bun.ptr.RefCount(@This(), "ref_count", deinit, .{});
cached_structure: PostgresCachedStructure = .{},
ref_count: RefCount = RefCount.init(),
fields: []protocol.FieldDescription = &[_]protocol.FieldDescription{},
parameters: []const int4 = &[_]int4{},
signature: Signature,
status: Status = Status.pending,
error_response: ?Error = null,
needs_duplicate_check: bool = true,
fields_flags: DataCell.Flags = .{},
pub const ref = RefCount.ref;
pub const deref = RefCount.deref;

pub const Error = union(enum) {
    protocol: protocol.ErrorResponse,
    postgres_error: AnyPostgresError,

    pub fn deinit(this: *@This()) void {
        switch (this.*) {
            .protocol => |*err| err.deinit(),
            .postgres_error => {},
        }
    }

    pub fn toJS(this: *const @This(), globalObject: *jsc.JSGlobalObject) JSError!JSValue {
        return switch (this.*) {
            .protocol => |err| err.toJS(globalObject),
            .postgres_error => |err| postgresErrorToJS(globalObject, null, err),
        };
    }
};

pub const Status = enum {
    pending,
    parsing,
    prepared,
    failed,

    pub fn isRunning(this: @This()) bool {
        return this == .parsing;
    }
};

pub fn checkForDuplicateFields(this: *PostgresSQLStatement) void {
    if (!this.needs_duplicate_check) return;
    this.needs_duplicate_check = false;

    var seen_numbers = std.array_list.Managed(u32).init(bun.default_allocator);
    defer seen_numbers.deinit();
    var seen_fields = bun.StringHashMap(void).init(bun.default_allocator);
    bun.handleOom(seen_fields.ensureUnusedCapacity(@intCast(this.fields.len)));
    defer seen_fields.deinit();

    // iterate backwards
    var remaining = this.fields.len;
    var flags: DataCell.Flags = .{};
    while (remaining > 0) {
        remaining -= 1;
        const field: *protocol.FieldDescription = &this.fields[remaining];
        switch (field.name_or_index) {
            .name => |*name| {
                const seen = seen_fields.getOrPut(name.slice()) catch unreachable;
                if (seen.found_existing) {
                    field.name_or_index = .duplicate;
                    flags.has_duplicate_columns = true;
                }

                flags.has_named_columns = true;
            },
            .index => |index| {
                if (std.mem.indexOfScalar(u32, seen_numbers.items, index) != null) {
                    field.name_or_index = .duplicate;
                    flags.has_duplicate_columns = true;
                } else {
                    bun.handleOom(seen_numbers.append(index));
                }

                flags.has_indexed_columns = true;
            },
            .duplicate => {
                flags.has_duplicate_columns = true;
            },
        }
    }

    this.fields_flags = flags;
}

pub fn deinit(this: *PostgresSQLStatement) void {
    debug("PostgresSQLStatement deinit", .{});

    this.ref_count.assertNoRefs();

    for (this.fields) |*field| {
        field.deinit();
    }
    bun.default_allocator.free(this.fields);
    bun.default_allocator.free(this.parameters);
    this.cached_structure.deinit();
    if (this.error_response) |err| {
        this.error_response = null;
        var _error = err;
        _error.deinit();
    }
    this.signature.deinit();
    bun.default_allocator.destroy(this);
}

pub fn structure(this: *PostgresSQLStatement, owner: JSValue, globalObject: *jsc.JSGlobalObject) PostgresCachedStructure {
    if (this.cached_structure.has()) {
        return this.cached_structure;
    }
    this.checkForDuplicateFields();

    // lets avoid most allocations
    var stack_ids: [70]jsc.JSObject.ExternColumnIdentifier = undefined;
    // lets de duplicate the fields early
    var nonDuplicatedCount = this.fields.len;
    for (this.fields) |*field| {
        if (field.name_or_index == .duplicate) {
            nonDuplicatedCount -= 1;
        }
    }
    const ids = if (nonDuplicatedCount <= jsc.JSObject.maxInlineCapacity()) stack_ids[0..nonDuplicatedCount] else bun.handleOom(bun.default_allocator.alloc(jsc.JSObject.ExternColumnIdentifier, nonDuplicatedCount));

    var i: usize = 0;
    for (this.fields) |*field| {
        if (field.name_or_index == .duplicate) continue;

        var id: *jsc.JSObject.ExternColumnIdentifier = &ids[i];
        switch (field.name_or_index) {
            .name => |name| {
                id.value.name = String.createAtomIfPossible(name.slice());
            },
            .index => |index| {
                id.value.index = index;
            },
            .duplicate => unreachable,
        }
        id.tag = switch (field.name_or_index) {
            .name => 2,
            .index => 1,
            .duplicate => 0,
        };
        i += 1;
    }

    if (nonDuplicatedCount > jsc.JSObject.maxInlineCapacity()) {
        this.cached_structure.set(globalObject, null, ids);
    } else {
        this.cached_structure.set(globalObject, jsc.JSObject.createStructure(
            globalObject,
            owner,
            @truncate(ids.len),
            ids.ptr,
        ), null);
    }

    return this.cached_structure;
}

const debug = bun.Output.scoped(.Postgres, .visible);

const PostgresCachedStructure = @import("../shared/CachedStructure.zig");
const Signature = @import("./Signature.zig");
const protocol = @import("./PostgresProtocol.zig");
const std = @import("std");
const DataCell = @import("./DataCell.zig").SQLDataCell;

const AnyPostgresError = @import("./AnyPostgresError.zig").AnyPostgresError;
const postgresErrorToJS = @import("./AnyPostgresError.zig").postgresErrorToJS;

const types = @import("./PostgresTypes.zig");
const int4 = types.int4;

const bun = @import("bun");
const JSError = bun.JSError;
const String = bun.String;

const jsc = bun.jsc;
const JSValue = jsc.JSValue;

File: src/sql_jsc/postgres/PostgresSQLStatement.rs
use core::cell::Cell;

use crate::jsc::{JSGlobalObject, JSValue, JsCell, JsResult};

use crate::postgres::error_jsc::postgres_error_to_js;
use crate::postgres::signature::Signature;
use crate::shared::cached_structure::CachedStructure as PostgresCachedStructure;
use crate::shared::sql_data_cell::{Flags as DataCellFlags, dedupe_columns};

use bun_sql::postgres::any_postgres_error::AnyPostgresError;
use bun_sql::postgres::postgres_protocol as protocol;
use bun_sql::postgres::postgres_types::int4;

bun_core::declare_scope!(Postgres, visible);

// Intrusive single-thread refcount (`CellRefCounted`). Shared between the owning query and the connection's prepared-statement map
// (each holds a `RefPtr`), so every field written after construction is
// interior-mutable and the statement is only ever reached as `&Self`.
#[derive(bun_ptr::CellRefCounted)]
pub struct PostgresSQLStatement {
    pub(crate) cached_structure: JsCell<PostgresCachedStructure>,
    ref_count: Cell<u32>,
    pub(crate) fields: JsCell<Vec<protocol::FieldDescription>>,
    pub(crate) parameters: JsCell<Box<[int4]>>,
    pub(crate) signature: Signature,
    pub(crate) status: Cell<Status>,
    pub(crate) error_response: JsCell<Option<Error>>,
    pub(crate) needs_duplicate_check: Cell<bool>,
    pub(crate) fields_flags: Cell<DataCellFlags>,
}

pub enum Error {
    Protocol(protocol::ErrorResponse),
    PostgresError(AnyPostgresError),
}

impl Error {
    // Cleanup is handled by `Drop` on
    // `protocol::ErrorResponse`, so no explicit `Drop` impl is needed here.

    pub(crate) fn to_js(&self, global_object: &JSGlobalObject) -> JsResult<JSValue> {
        match self {
            Error::Protocol(err) => Ok(crate::postgres::protocol::error_response_jsc::to_js(
                err,
                global_object,
            )),
            Error::PostgresError(err) => Ok(postgres_error_to_js(global_object, None, *err)),
        }
    }
}

pub use bun_sql::shared::statement_status::Status;

impl PostgresSQLStatement {
    /// A new statement with one ref, owned by the returned handle.
    pub(crate) fn new(signature: Signature, status: Status) -> bun_ptr::RefPtr<Self> {
        bun_ptr::RefPtr::new(Self {
            cached_structure: JsCell::new(PostgresCachedStructure::default()),
            ref_count: Cell::new(1),
            fields: JsCell::new(Vec::new()),
            parameters: JsCell::new(Box::default()),
            signature,
            status: Cell::new(status),
            error_response: JsCell::new(None),
            needs_duplicate_check: Cell::new(true),
            fields_flags: Cell::new(DataCellFlags::default()),
        })
    }

    /// The stored error as a JS value (`undefined` if none is stored).
    pub(crate) fn error_response_to_js(&self, global_object: &JSGlobalObject) -> JsResult<JSValue> {
        match self.error_response.get() {
            Some(e) => e.to_js(global_object),
            None => Ok(JSValue::UNDEFINED),
        }
    }

    /// Record `err` and mark the statement failed.
    pub(crate) fn fail(&self, err: Error) {
        self.status.set(Status::Failed);
        self.error_response.set(Some(err));
    }

    pub(crate) fn check_for_duplicate_fields(&self) {
        if !self.needs_duplicate_check.get() {
            return;
        }
        self.needs_duplicate_check.set(false);

        let flags = self.fields.with_mut(|fields| {
            dedupe_columns(fields.iter_mut().rev().map(|f| &mut f.name_or_index))
        });
        self.fields_flags.set(flags);
    }

    /// The cached JSC structure for this statement's columns, building it on
    /// first use.
    pub(crate) fn structure(
        &self,
        owner: JSValue,
        global_object: &JSGlobalObject,
    ) -> &PostgresCachedStructure {
        if !self.cached_structure.get().has() {
            self.check_for_duplicate_fields();
            let fields = self.fields.get();
            self.cached_structure.with_mut(|cs| {
                cs.build_from_columns(
                    global_object,
                    owner,
                    fields.iter().map(|f| &f.name_or_index),
                )
            });
        }
        self.cached_structure.get()
    }
}

impl Drop for PostgresSQLStatement {
    fn drop(&mut self) {
        bun_core::scoped_log!(Postgres, "PostgresSQLStatement deinit");

        debug_assert_eq!(self.ref_count.get(), 0, "ref_count.assertNoRefs()");

        // `fields` (Vec<FieldDescription>): each element's Drop runs, then the buffer frees.
        // `parameters` (Box<[int4]>): freed by Drop.
        // `cached_structure`, `error_response`, `signature`: Drop.
    }
}
